nifi-nar-bundles/nifi-standard-services/nifi-oauth2-provider-bundle/nifi-oauth2-provider-service/src/main/java/org/apache/nifi/oauth2/StandardOauth2AccessTokenProvider.java

```diff
@@ -0,0 +1,300 @@
+
+    private volatile AccessToken accessDetails;
+
+    @Override
+    public List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return PROPERTIES;
+    }
+
+    @OnEnabled
+    public void onEnabled(ConfigurationContext context) {
+        authorizationServerUrl = context.getProperty(AUTHORIZATION_SERVER_URL).evaluateAttributeExpressions().getValue();
+
+        httpClient = createHttpClient(context);
+
+        grantType = context.getProperty(GRANT_TYPE).getValue();
+        username = context.getProperty(USERNAME).evaluateAttributeExpressions().getValue();
+        password = context.getProperty(PASSWORD).getValue();
+        clientId = context.getProperty(CLIENT_ID).evaluateAttributeExpressions().getValue();
+        clientSecret = context.getProperty(CLIENT_SECRET).getValue();
+    }
+
+    @OnDisabled
+    public void onDisabled() {
+        accessDetails = null;
+    }
+
+    @Override
+    protected Collection<ValidationResult> customValidate(ValidationContext validationContext) {
+        final List<ValidationResult> validationResults = new ArrayList<>(super.customValidate(validationContext));
+
+        if (
+            validationContext.getProperty(GRANT_TYPE).getValue().equals(CLIENT_CREDENTIALS_GRANT_TYPE.getValue())
+                && !validationContext.getProperty(CLIENT_ID).isSet()
+        ) {
+            validationResults.add(new ValidationResult.Builder().subject(CLIENT_ID.getDisplayName())
+                .valid(false)
+                .explanation(String.format(
+                    "When '%s' is set to '%s', '%s' is required",
+                    GRANT_TYPE.getDisplayName(),
+                    CLIENT_CREDENTIALS_GRANT_TYPE.getDisplayName(),
+                    CLIENT_ID.getDisplayName())
+                )
+                .build());
+        }
+
+        return validationResults;
+    }
+
+    protected OkHttpClient createHttpClient(ConfigurationContext context) {
+        OkHttpClient.Builder clientBuilder = new OkHttpClient.Builder();
+
+        SSLContextService sslService = context.getProperty(SSL_CONTEXT).asControllerService(SSLContextService.class);
+        if (sslService != null) {
+            final X509TrustManager trustManager = sslService.createTrustManager();
+            SSLContext sslContext = sslService.createContext();
+            clientBuilder.sslSocketFactory(sslContext.getSocketFactory(), trustManager);
+        }
+
+        return clientBuilder.build();
+    }
+
+    @Override
+    public AccessToken getAccessDetails() {
+        if (this.accessDetails == null) {
+            acquireAccessDetails();
+        } else if (this.accessDetails.isExpired()) {
+            if (this.accessDetails.getRefreshToken() == null) {
+                acquireAccessDetails();
+            } else {
+                try {
+                    refreshAccessDetails();
+                } catch (Exception e) {
+                    getLogger().info("Couldn't refresh access token.", e);
```

Comment:
@exceptionfactory , although I don't have a strong opinion on this particular point my concern with `null`ing the `accessDetails` is that it might be a bit confusing to the reader and we don't save too much here:
- if there'll be multiple calls to this method and the refresh-acquire pair fails then there's some bigger problem -> the "extra" refresh try doesn't make too much difference.
- But if the subsequent call succeeds at `acquireAccessDetails` then we only have one unnecessary refresh call.

Compared to how confusing the `accessDetails = null` could be it might not be worth it.
WDYT?

Reply:
I'd rather leave out this type of micromanagement code. The benefit doesn't translate to practical advantage in real-life situations. Also, retrying the refresh after the new token acquisition failed very well may make sense and could succeed. Maybe it fails more often than not but the point is, it doesn't really matter.

Reply:
I agree that setting `accessDetails` to `null` could be confusing. That is part of the reason for recommend more discrete exception handling. Here is the potential scenario:

1. Initial `acquireAccessDetails()` succeeds and includes a refresh token
2. Access Token expires, leading to the service calling `refreshAccessDetails()`
3. The call to `refreshAccessDetails()` fails due to an `IOException`
4. The call to `acquireAccessDetails()` fails due to an `IOException`

As long there are continuing connection issues, the service will continue calling both `refreshAccessDetails()` and `acquireAccessDetails()` on every `getAccessDetails()` invocation. The same thing would happen in the case of both methods throwing some type of `RuntimeException` related to a bad HTTP response. Logs might be enough to identify the problem, but the main concern is the state of the access details.  Given that this new service implementation intends to hide the complexities of the state from the caller, it seems important to have the implementation itself as clear as possible.

Reply:
I was thinking a bit more about this.
I understand the reason of setting the `accessDetails` to `null`. What if the `acquireAccessDetails` resets it at the beginning? Logically it belongs there and it's cleaner for the reader that after the `acquireAccessDetails` it either has a value or is `null` due to an exception.

Reply:
Thanks @adenes, I was thinking about that approach as well. Setting `accessDetails` to `null` in `acquireAccessDetails` does seem like the logical location to address this concern.

Reply:
Would it be possible to move past theoretical potential scenarios and consider reproducible issues?

I understand how the mentioned scenario would work. I thought about it already a long time ago.
Yes, it's possible that there would be multiple unsuccessful calls to the access server.
Yes, if the access server is unaccessible, we would have _2_ unsuccessful calls at each attempt.

Where I disagree is that this is actually a problem. Would we be missing any important detail to be able to solve the issue? Would there be a scenario where this implementation would not work but another implementation would? I see no proof for any of those to be true.

As for setting `acquireAccessDetails` to `null`, again I see no case where this would be better than leaving it as it is. In fact I'd argue that in case we would like to do a memory dump it's better to keep the last working `accessDetails` all the time.

Reply:
Thanks for the reply @tpalfy. Although runtime issues can create some strange edge cases, I can appreciate not wanting to get stuck on theoretical concerns.

In order to move this forward, creating unit tests that exercise exception conditions would satisfy my concerns. It would be very helpful to have one unit test that throws some type of IOException, and another that throws an exception as the result of a non-success HTTP status. Having those two tests would exercise this block of code and cover the bases.

With the addition of unit tests, and if @MikeThomsen signs off on this refactored approach, I can support these changes.

Reply:
Expecting better test coverage is fair enough. Going to add ones that cover the various error cases.

To be honest previously I thought about adding unit tests but the non-error cases would rely so much on mocks that it would no longer ensure proper functionality in a real-life environment. So I settled for manual testing. But for the error-cases it makes sense to add some unit tests to cover the error handling.